lib/src/dag_walk.rs

```diff
@@ -487,16 +487,31 @@ where
     NI: IntoIterator<Item = Result<T, E>>,
 {
     let start: Vec<T> = start.into_iter().try_collect()?;
-    let mut reachable: HashSet<T> = start.iter().cloned().collect();
-    dfs_ok(start.into_iter().map(Ok), id_fn, |node| {
-        let neighbors: Vec<Result<T, E>> = neighbors_fn(node).into_iter().collect();
-        for neighbor in neighbors.iter().filter_map(|x| x.as_ref().ok()) {
-            reachable.remove(neighbor);
+    let mut heads: HashSet<T> = start.iter().cloned().collect();
+    // Do a BFS until we have only one item left in the frontier. That frontier must
+    // have originated from one of the heads, and since there can't be cycles,
+    // it won't be able to eliminate any other heads.
+    let mut frontier: HashSet<T> = heads.clone();
+    let mut visited: HashSet<ID> = HashSet::new();
+    let mut root_reached = false;
+    while frontier.len() > 1 || (!frontier.is_empty() && root_reached) {
+        visited.extend(frontier.iter().map(&id_fn));
+        frontier = frontier
+            .iter()
+            .flat_map(|node| {
+                let neighbors = neighbors_fn(node).into_iter().collect_vec();
+                if neighbors.is_empty() {
+                    root_reached = true;
+                }
+                neighbors
+            })
+            .try_collect()?;
+        for node in &frontier {
+            heads.remove(node);
         }
-        neighbors
-    })
-    .try_for_each(|r| r.map(|_| ()))?;
-    Ok(reachable)
+        frontier.retain(|node| !visited.contains(&id_fn(node)));
```

Comment:
nit: maybe `visited` can be updated here?
```rust
frontier.retain(|node| visited.insert(id_fn(node)));
```
In BFS, I think it's common to deduplicate when queueing.

Reply:
Good idea. That avoids calling `id_fn()` twice.